### ISSUE TITLE
Removing asn1_len from the structure in ocsp_ht.c / http_client.c broke the nbio state machine

### ISSUE THREAD

**Issue by jwalch** [openssl/openssl#13938]
Another unintended consequence of https://github.com/openssl/openssl/commit/29f178bddfdbd11218fbcba0b8060297696968e3

The issue specifically arises when we end up hitting the BIO_should_retry() check in the OHS_WRITE state (ie, we return -1).

```
int OSSL_HTTP_REQ_CTX_nbio(OSSL_HTTP_REQ_CTX *rctx)
{
    int i;
    long n, n_to_send = 0;
[...]
    case OHS_WRITE_INIT:
        n_to_send = BIO_get_mem_data(rctx->mem, NULL);
        rctx->state = OHS_WRITE;

        /* fall thru */
    case OHS_WRITE:
        n = BIO_get_mem_data(rctx->mem, &p);

        i = BIO_write(rctx->wbio, p + (n - n_to_send), n_to_send);

        if (i <= 0) {
            if (BIO_should_retry(rctx->wbio))
                return -1;
[...]
```

It used to be that the asn1_len struct member would "remember" the value from the OHS_WRITE_INIT state on subsequent calls. So if the BIO_write() in OHS_WRITE failed for a retry-able reason, when the caller did in fact retry the third parameter to said BIO_write() call would remain the same. However, after the refactoring, that's not the case. Instead, when we retry the n_to_send local variable is initialized to 0 again and since we go straight into OHS_WRITE again (not OHS_WRITE_INIT) it's never set to the "right" value.

EDIT: Removed some proposed solution text. Discussion of why it was insufficient is in the linked PR if anyone wants it.

**Comment by mattcaswell** [comment]
@DDvO

**Comment by DDvO** [comment]
Fix ist #13960 (commit de84ccf0f2e019bffc8012a926f5bd183d15fa94).
